### PR TITLE
(maint) Compare Environment#name in compiler

### PR DESCRIPTION
**Pull request by justinstoller**
During catalog compilation, after loading the requested environment we load the node and server specified environment. Previously, We would then compare the Environment objects to determine if the requested environment matched the server specified environment. However, if the environment cache is flushed between those two environments being loaded they could be different object instances representing the same desired environment.

To resolve this race condition we now check the environment names (Environment#name canoncalizes to symbols to avoid symbol/string mismatch).

### CODE REVIEW

**Comment by justinstoller** [comment]
This is an attempt to resolve support request PE-36986 , a quick review would be appreciated but I want to wait until we hear back from the customer before merging.

**Comment by joshcooper** [comment]
> However, if the environment cache is flushed between those two environments being loaded they could be different object instances representing the same desired environment.

In theory, the `Puppet::Node::Environment` class overrides `==`, `eql?` and `hash` https://github.com/puppetlabs/puppet/blob/e177633009491a3443b5734568a5238fc83da19f/lib/puppet/node/environment.rb#L538-L549

So it should be possible to check if two environment objects have the same value. But the check depends on the `manifest` and `full_modulepath`. When the env cache is flushed and the environment is reloaded, does it have a different `manifest` or `modulepath` due to versioned code symlinks?

That said, I think comparing environment names is correct, since all we care about is whether the agent's requested environment matches the server-specified environment (based on classification rules).

Tangent 1: It seems strange that we're using `kind_of?` instead of `instance_of?` as the former means you can compare remote environment with a non-remote one.

```
irb(main):016:0> a = Puppet::Node::Environment.create(:production, [])
=> <Puppet::Node::Environment:442580 @name="production" @manifest="no_manifest" @modulepath="" >
irb(main):017:0> b = Puppet::Node::Environment.remote(:production)
=> <Puppet::Node::Environment::Remote:458740 @name="production" @manifest="no_manifest" @modulepath="" >
irb(main):018:0> a == b
=> true
```

We're probably relying on that in various places, but it seems like a footgun.

Tangent 2: In theory, cached environment names should always be symbols, but it wasn't always the case https://puppet.atlassian.net/browse/PUP-10955.

**Comment by justinstoller** [comment]
@joshcooper great callout about overriding `==`, I assume we should not alter that implementation and in this case explicitly compare names. Do you think so? If so, I'll add some of your findings to the comments.

**Comment by joshcooper** [comment]
> I assume we should not alter that implementation and in this case explicitly compare names

Yeah I'm pretty something is relying on being able to compare real `Environment` and fake `Environment::Remote` objects, probably in the configurer. So I'm :+1: to comparing names, since in this case we only care if the agent requested a different environment **name** than the server's node terminus said to use for compilation.

**Comment by justinstoller** [comment]
Heard back from @vchepkov and it looks like this is working so far. Even if it does turn out not to be the solution to his problem I think it is still a legit bug that should be fixed and doesn't appear to cause regressions. I've updated the comment and commit message based on Josh's comments. I think this is ready to go now.

**Comment by joshcooper** [comment]
LGTM, could you [add/update a test around here](https://github.com/puppetlabs/puppet/blob/298762f24217e057147194cf95525cb4c40ffb54/spec/unit/indirector/catalog/compiler_spec.rb#L274-L280) so that different `Environment` objects (with the same name) are set on the `node` and `request`? I think it's most realistic to set a remote environment (created via  `Puppet::Node::Environment.remote(name)`) on the `request`.

**Comment by justinstoller** [comment]
Thanks, Josh! I've added a test now.

**Comment by vchepkov** [comment]
FYI, running this patch since October, 26th, issues with empty catalog hasn't reoccured since

**Comment by justinstoller** [comment]
I think this is waiting on an approval from @puppetlabs/phoenix , let me know if there's something else I can or should do!

**Review by joshcooper — APPROVED** [review]

**Pull request merged**
